src/dmd/backend/cgcse.d

```diff
@@ -0,0 +1,121 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Manage the memory allocated on the runtime stack to save Common Subexpressions (CSE).
+ *
+ * Copyright:   Copyright (C) 1985-1998 by Symantec
+ *              Copyright (C) 2000-2019 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/backend/cgcse.d, backend/cgcse.d)
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/backend/cgcse.d
+ */
+
+module dmd.backend.cgcse;
```

Comment:
I suggest you take the opportunity to give this new module a better less cryptic name for those not familiar with the code base. 

BTW, all these `cg` prefixes are a good indication that a package would be better.

Reply:
CSE is commonplace compiler jargon. If you still need to know what it is, read the header comment in the file.

Reply:
True, but `cgcse` looks like someone had a seizure at the keyboard. Also CSE is an optimisation pass and has nothing to do with codegen, which is what I associate with `cg`.

Reply:
> CSE is an optimisation pass and has nothing to do with codegen

As you can see the codegen is very much involved with CSE.

Reply:
The code doesn't look like it does much at all. 
CSE is usually an IR to IR pass, I don't know the backend is doing if its not that.

Reply:
> The code doesn't look like it does much at all.

That's right, for a first step.

_For easy review, I kept this code movement as cut/paste. It's really hard to review such if it is all refactored as well as being moved. As I explain again and again, small incremental PRs are far easier to review than large ones, and far easier to debug if a regression is introduced. In light of that, this should be reviewed as progress towards a goal, not whether the goal is achieved._

> CSE is usually an IR to IR pass

Part of it is, see gdag.d The rest is in the code generator.